platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileWriter.java

```diff
@@ -163,64 +238,82 @@ public synchronized long writeCopiedDataItem(final long serializedVersion, final
     /**
      * Store data item in file returning location it was stored at.
      *
-     * @param dataItem
-     * 		the data item to write
+     * @param dataItem the data item to write
      * @return the data location of written data in bytes
-     * @throws IOException
-     * 		if there was a problem appending data to file
+     * @throws IOException if there was a problem appending data to file
      */
     public synchronized long storeDataItem(final D dataItem) throws IOException {
-        // find offset for the start of this new data item, we assume we always write data in a whole number of blocks
-        final long byteOffset = writePosition;
+        // find offset for the start of this new data item, we assume we always write data in a
+        // whole number of blocks
+        final int currentWritingMmapPos = writingMmap.position();
+        final long byteOffset = mmapPositionInFile + currentWritingMmapPos;
         // write serialized data
-        final int totalDataWritten = dataItemSerializer.serialize(dataItem, writingStream);
-        // update write position
-        writePosition += totalDataWritten;
+        try {
+            dataItemSerializer.serialize(dataItem, writingMmap);
+        } catch (final BufferOverflowException e) {
+            // Buffer overflow indicates we need to remap the current writing buffer and
+            // retry to serialize the data item
```

Comment:
The similar code above sets position between remap and retry, should this block do the same?

Reply:
No, serialization is stateless, no state to reset before retry. The code above is slightly different: it copies a data item (as a byte buffer) from another file. When retried, the buffer must be reset to a previous state (position).